vulnerabilities/models.py

```diff
@@ -642,6 +644,152 @@ def get_absolute_url(self):
+        """
+        Return a list of `packages` sorted by version.
+        """
+        if not packages:
+            return []
+
+        version_class = RANGE_CLASS_BY_SCHEMES[packages[0].type].version_class
+        return sorted(
+            packages,
+            key=lambda x: version_class(x.version),
+        )
+
+    @property
+    def version_class(self):
+        return RANGE_CLASS_BY_SCHEMES[self.type].version_class
+
+    @property
+    def current_version(self):
+        return self.version_class(self.version)
+
+    @property
+    def fixed_package_details(self):
+        """
+        Return a mapping of vulnerabilities that affect this package and the next and
+        latest non-vulnerable versions.
+        """
+        package_details = {}
+        package_details["purl"] = PackageURL.from_string(self.purl)
+
+        next_non_vulnerable, latest_non_vulnerable = self.get_non_vulnerable_versions()
+        package_details["next_non_vulnerable"] = next_non_vulnerable
+        package_details["latest_non_vulnerable"] = latest_non_vulnerable
+
+        package_details["vulnerabilities"] = self.get_affecting_vulnerabilities()
+
+        return package_details
+
+    def get_non_vulnerable_versions(self):
+        """
+        Return a tuple of the next and latest non-vulnerable versions as PackageURLs.  Return a tuple of
+        (None, None) if there is no non-vulnerable version.
+        """
+        package_versions = self.get_fixed_by_package_versions(fix=False)
+
+        non_vulnerable_versions = []
+        for version in package_versions:
+            if not version.is_vulnerable:
+                non_vulnerable_versions.append(version)
+
+        later_non_vulnerable_versions = []
+        for non_vuln_ver in non_vulnerable_versions:
+            if self.version_class(non_vuln_ver.version) > self.current_version:
+                later_non_vulnerable_versions.append(non_vuln_ver)
+
+        if later_non_vulnerable_versions:
+            sorted_versions = self.sort_by_version(later_non_vulnerable_versions)
+            next_non_vulnerable_version = sorted_versions[0]
+            latest_non_vulnerable_version = sorted_versions[-1]
+
+            next_non_vulnerable = PackageURL.from_string(next_non_vulnerable_version.purl)
+            latest_non_vulnerable = PackageURL.from_string(latest_non_vulnerable_version.purl)
+
+            return next_non_vulnerable, latest_non_vulnerable
+
+        return None, None
+
+    def get_affecting_vulnerabilities(self):
+        """
+        Return a list of vulnerabilities that affect this package together with information regarding
+        the versions that fix the vulnerabilities.
+        """
+        package_details_vulns = []
+
+        fixed_by_packages = self.get_fixed_by_package_versions(fix=True)
+
+        package_vulnerabilities = self.vulnerabilities.filter(
+            packagerelatedvulnerability__fix=False
+        ).prefetch_related(
```

Comment:
This filter exists as a `affected()` queryset method. I would recommend using it to improve readability.

https://github.com/nexB/vulnerablecode/blob/e42d09567c014f789d8af2ed166c61e04138da9d/vulnerabilities/models.py#L434

Reply:
@tdruez I've tried a number of approaches to this comment, including those below, but I think I just do not understand how to implement your comment.

```
        # package_vulnerabilities = self.vulnerabilities.filter(
        #     packagerelatedvulnerability__fix=False
        # ).prefetch_related(
        #     Prefetch(
        #         "packages",
        #         queryset=fixed_by_packages,
        #         to_attr="fixed_packages",
        #     )
        # )

        # This returns packages affected by a vulnerability.
        # package_vulnerabilities = Package.objects.affected()
        # package_vulnerabilities.prefetch_related(
        #     Prefetch(
        #         "packages",
        #         queryset=fixed_by_packages,
        #         to_attr="fixed_packages",
        #     )
        # )

        package_vulnerabilities = Package.objects.affected().prefetch_related(
            Prefetch(
                "packages",
                queryset=fixed_by_packages,
                to_attr="fixed_packages",
            )
        )
```

And I also remain unclear on how to implement your comment below re converting the Package method `get_fixed_by_package_versions()` to a `PackageQuerySet` method.  Any hints or examples you can point me to?  Or have I spent too much time already on these?

Reply:
From https://github.com/nexB/vulnerablecode/blob/e42d09567c014f789d8af2ed166c61e04138da9d/vulnerabilities/models.py#L434

```
def affected(self):
        """
        Return only packages affected by a vulnerability.
        """
        return self.filter(packagerelatedvulnerability__fix=False)
```

```
# Not DRY and less readable
package_vulnerabilities = self.vulnerabilities.filter(packagerelatedvulnerability__fix=False)

# Re-use of existing QuerySet method and much simpler to read
package_vulnerabilities = self.vulnerabilities.affected()
```

As an extra note on this from your examples:
```
# This returns packages affected by a vulnerability.
        # package_vulnerabilities = Package.objects.affected()
        # package_vulnerabilities.prefetch_related(
        #     Prefetch(
        #         "packages",
        #         queryset=fixed_by_packages,
        #         to_attr="fixed_packages",
        #     )
        # )
```

Here, you do not store the results of applying `prefetch_related`, wich make this statement useless, while still trigerring database queries.

Reply:
Thanks @tdruez .  I tried your suggested approach yesterday, with the prefetch_related, and just ran it again now, getting 14 failed tests.

```
        package_vulnerabilities = self.vulnerabilities.affected().prefetch_related(
            Prefetch(
                "packages",
                queryset=fixed_by_packages,
                to_attr="fixed_packages",
            )
        )
```

results:

```
FAILED vulnerabilities/tests/test_api.py::APITestCasePackage::test_api_response - AttributeError: 'ManyRelatedManager' object has no attribute 'affected'
FAILED vulnerabilities/tests/test_api.py::APITestCasePackage::test_api_status - AttributeError: 'ManyRelatedManager' object has no attribute 'affected'
FAILED vulnerabilities/tests/test_api.py::APITestCasePackage::test_api_with_ignorning_qualifiers - AttributeError: 'ManyRelatedManager' object has no attribute 'affected'
FAILED vulnerabilities/tests/test_api.py::APITestCasePackage::test_api_with_lesser_and_greater_fixed_by_packages - AttributeError: 'ManyRelatedManager' object has no attribute 'affected'
FAILED vulnerabilities/tests/test_api.py::APITestCasePackage::test_api_with_namespace_filter - AttributeError: 'ManyRelatedManager' object has no attribute 'affected'
FAILED vulnerabilities/tests/test_api.py::BulkSearchAPIPackage::test_bulk_api_response - AttributeError: 'ManyRelatedManager' object has no attribute 'affected'
FAILED vulnerabilities/tests/test_api.py::BulkSearchAPIPackage::test_bulk_api_response_with_ignoring_qualifiers - AttributeError: 'ManyRelatedManager' object has no attribute 'affected'
FAILED vulnerabilities/tests/test_api.py::BulkSearchAPIPackage::test_bulk_api_response_with_ignoring_subpath - AttributeError: 'ManyRelatedManager' object has no attribute 'affected'
FAILED vulnerabilities/tests/test_api.py::TestDebianResponse::test_query_by_name - AttributeError: 'ManyRelatedManager' object has no attribute 'affected'
FAILED vulnerabilities/tests/test_api.py::TestDebianResponse::test_query_by_package_url_without_namespace - AttributeError: 'ManyRelatedManager' object has no attribute 'affected'
FAILED vulnerabilities/tests/test_api.py::TestDebianResponse::test_query_qualifier_filtering - AttributeError: 'ManyRelatedManager' object has no attribute 'affected'
FAILED vulnerabilities/tests/test_api.py::TestSerializers::test_package_serializer - AttributeError: 'ManyRelatedManager' object has no attribute 'affected'
FAILED vulnerabilities/tests/test_api.py::APITestCaseVulnerability::test_api_with_single_vulnerability - AttributeError: 'ManyRelatedManager' object has no attribute 'affected'
FAILED vulnerabilities/tests/test_api.py::APITestCaseVulnerability::test_api_with_single_vulnerability_with_filters - AttributeError: 'ManyRelatedManager' object has no attribute 'affected'
```

and trying your approach without the prefetch_related, 14 different failures

```
package_vulnerabilities = self.vulnerabilities.affected()
```

results:

```
FAILED vulnerabilities/tests/test_api.py::APITestCasePackage::test_api_response - AttributeError: 'ManyRelatedManager' object has no attribute 'affected'
FAILED vulnerabilities/tests/test_api.py::APITestCasePackage::test_api_status - AttributeError: 'ManyRelatedManager' object has no attribute 'affected'
FAILED vulnerabilities/tests/test_api.py::APITestCasePackage::test_api_with_ignorning_qualifiers - AttributeError: 'ManyRelatedManager' object has no attribute 'affected'
FAILED vulnerabilities/tests/test_api.py::APITestCasePackage::test_api_with_lesser_and_greater_fixed_by_packages - AttributeError: 'ManyRelatedManager' object has no attribute 'affected'
FAILED vulnerabilities/tests/test_api.py::APITestCasePackage::test_api_with_namespace_filter - AttributeError: 'ManyRelatedManager' object has no attribute 'affected'
FAILED vulnerabilities/tests/test_api.py::BulkSearchAPIPackage::test_bulk_api_response - AttributeError: 'ManyRelatedManager' object has no attribute 'affected'
FAILED vulnerabilities/tests/test_api.py::BulkSearchAPIPackage::test_bulk_api_response_with_ignoring_qualifiers - AttributeError: 'ManyRelatedManager' object has no attribute 'affected'
FAILED vulnerabilities/tests/test_api.py::BulkSearchAPIPackage::test_bulk_api_response_with_ignoring_subpath - AttributeError: 'ManyRelatedManager' object has no attribute 'affected'
FAILED vulnerabilities/tests/test_api.py::TestDebianResponse::test_query_by_name - AttributeError: 'ManyRelatedManager' object has no attribute 'affected'
FAILED vulnerabilities/tests/test_api.py::TestDebianResponse::test_query_by_package_url_without_namespace - AttributeError: 'ManyRelatedManager' object has no attribute 'affected'
FAILED vulnerabilities/tests/test_api.py::TestDebianResponse::test_query_qualifier_filtering - AttributeError: 'ManyRelatedManager' object has no attribute 'affected'
FAILED vulnerabilities/tests/test_api.py::TestSerializers::test_package_serializer - AttributeError: 'ManyRelatedManager' object has no attribute 'affected'
FAILED vulnerabilities/tests/test_api.py::APITestCaseVulnerability::test_api_with_single_vulnerability - AttributeError: 'ManyRelatedManager' object has no attribute 'affected'
FAILED vulnerabilities/tests/test_api.py::APITestCaseVulnerability::test_api_with_single_vulnerability_with_filters - AttributeError: 'ManyRelatedManager' object has no attribute 'affected'

```

So, I remain unclear about how to implement your comment.

Reply:
On closer inspection it looks like both approaches result in the same 14 failed tests and same errors.

Reply:
@johnmhoran my bad, `affected()` only exist on the `PackageQuerySet`, not on the `VulnerabilityQuerySet`, and here you're dealing with `self.vulnerabilities`, not packages.

You should implement the pendant of `PackageQuerySet.affected` onto the `VulnerabilityQuerySet` class, probably a slightly different function name as it's not exactly the same context, maybe `affecting`? Or you already have a name for this case.

Reply:
Hi @tdruez and thank you!  I was wondering yesterday why we were not using the vuln queryset manager.  Just now, I created a new method

```
class VulnerabilityQuerySet(BaseQuerySet):
    def affected_packages(self):
        """
        Return only packages affected by a vulnerability.
        """
        return self.filter(packagerelatedvulnerability__fix=False)

```

and in the Package class I replaced my existing code with this new `affected_packages` queryset method

```
        package_vulnerabilities = self.vulnerabilities.affected_packages().prefetch_related(
            Prefetch(
                "packages",
                queryset=fixed_by_packages,
                to_attr="fixed_packages",
            )
        )
```

and no test failures!

```
-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
=========================================================================== 307 passed, 1 skipped, 1 deselected, 9 warnings in 33.49s ===========================================================================

(venv) Wed Nov 22, 2023 08:55 AM  /home/jmh/dev/nexb/vulnerablecode jmh (1228-fixed-affected-version-matching)
$
```

Reply:
But I see there's already an `affected_packages()` inside the Vulnerability class itself so I'll need to rename my new queryset manager method....

Reply:
The VulnerabilityQuerySet is likely to return Vulnerability objects, so the `affected_packages()` is not great anyway.

Reply:
Renamed to `def affecting_vulnerabilities(self)`.